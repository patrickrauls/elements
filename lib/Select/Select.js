import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Flag from '../Flag';
import I from '../I';

class Select extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: props.defaultValue || '',
      valid: false,
      dirty: false,
      status: '',
      focus: false
    };
  }
  checkStatus() {
    if (this.state.valid && this.state.dirty) {
      this.setState({ status: 'valid' });
    }
    if (this.props.required && this.state.dirty && !this.state.value) {
      this.setState({ status: 'required' });
    } else if (this.state.dirty && !this.state.valid) {
      this.setState({ status: 'invalid' });
    }
  }
  handleChange = e => {
    if (this.props.onChange) {
      this.props.onChange(e);
    }
    const { value } = e.target;
    this.setState(
      {
        valid: e.target.checkValidity(),
        dirty: true,
        value
      },
      () => {
        this.checkStatus();
      }
    );
  };
  handleClick = ({ currentTarget }) => {
    const select = [...currentTarget.children].find(element => {
      return element.tagName === 'SELECT';
    });
    select.focus();
  };
  handleFocus = e => {
    if (this.props.onFocus) {
      this.props.onFocus(e);
    }
    this.setState({
      focus: true,
      dirty: true
    });
  };
  handleBlur = e => {
    if (this.props.onBlur) {
      this.props.onBlur(e);
    }
    this.setState(
      {
        focus: false
      },
      () => {
        this.checkStatus();
      }
    );
  };
  render() {
    const {
      className,
      icon,
      name,
      onChange,
      required,
      defaultValue,
      onBlur,
      onFocus,
      placeholder,
      disabled,
      children,
      ...rest
    } = this.props;
    return (
      <div
        className={`
          ${className}
          ellies-select-container
          ${this.state.focus ? 'ellies-select-focused' : ''}
          ellies-${this.state.status}
        `}
        onClick={this.handleClick}
        onKeyDown={this.handleClick}
        role="presentation"
      >
        {icon && (
          <I
            className="ellies-select-icon"
            icon={icon}
          />
        )}
        <select
          {...rest}
          className={`
          ${name}
          ${name}-select-element
          ${this.state.dirty ? '' : 'placeholder'}
          `}
          disabled={disabled}
          name={name}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onFocus={this.handleFocus}
          required={required}
          value={this.state.value}
        >
          {placeholder && (
            <option disabled value="">
              {placeholder}
            </option>
          )}
          {children}
        </select>
        <Flag type={this.state.status || 'hidden'} />
      </div>
    );
  }
}

Select.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element)
  ]),
  className: PropTypes.string,
  defaultValue: PropTypes.string,
  disabled: PropTypes.bool,
  icon: PropTypes.string,
  name: PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  placeholder: PropTypes.string,
  required: PropTypes.bool
};

Select.defaultProps = {
  children: undefined,
  className: '',
  defaultValue: undefined,
  disabled: false,
  icon: '',
  name: '',
  onBlur: undefined,
  onChange: undefined,
  onFocus: undefined,
  placeholder: undefined,
  required: false
};

export default Select;
