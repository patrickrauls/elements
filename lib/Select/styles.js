import { css } from 'styled-components';

const selectStyles = css`
  position: relative;
  margin-top: 2.1875rem;
  border-bottom: 0.125rem solid #474f4f;
  width: 100%;
  background-color: rgba(0, 0, 0, 0);
  padding: 0;
  opacity: ${props => props.disabled ? '0.4' : '1'};
  cursor: pointer;
  &.ellies-disabled-select {
    opacity: 0.4;
  }
  &.ellies-select-focused {
    border-bottom: 0.125rem solid #021828;
    color: #021828;
  }
  &.ellies-invalid,
  &.ellies-required {
    color: #af1a12;
    border-bottom: 0.125rem solid #af1a12;
  }
  select {
    cursor: pointer;
    -moz-appearance: none;
    -webkit-appearance: none;
    border: none;
    width: 100%;
    padding: 0 1.125rem 0.25rem;
    font-size: 1rem;
    background-color: rgba(0, 0, 0, 0);
  }
  select.placeholder {
    color: #474f4f;
    opacity: 0.9;
  }
  select:focus {
    outline: none;
    color: #021828;
  }
  .ellies-select-icon {
    position: absolute;
    height: 1rem;
    width: 0.875rem;
    left: 0;
    bottom: 0.25rem;
  }
  select.invalid:not(:focus) ~ .flag {
    display: block;
    animation: fadein 0.3s;
  }
  .select:focus {
    display: none;
  }
`;

export default selectStyles;
