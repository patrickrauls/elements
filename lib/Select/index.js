import styled from 'styled-components';
import selectStyles from './styles';
import Select from './Select';

const StyledSelect = styled(Select)`
  ${selectStyles};
`;

export default StyledSelect;
