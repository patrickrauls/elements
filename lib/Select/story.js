import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, select, boolean } from '@storybook/addon-knobs';
import Select from './index';

storiesOf('Select', module)
  // .addDecorator(withTheme)
  .addDecorator(withKnobs)
  .add('one child', () => (
    <Select
      disabled={boolean('disabled', false)}
      invalid={boolean('invalid', false)}
      name={select('name', ['default', 'timezone', 'timeslot'])}
      onChange={action('changed')}
    >
      <option>Hey</option>
    </Select>
  ))
  .add('with placeholder', () => (
    <Select
      disabled={boolean('disabled', false)}
      placeholder="none"
      invalid={boolean('invalid', false)}
      name={select('name', ['default', 'timezone', 'timeslot'])}
      onChange={action('changed')}
    >
      <option selected>select</option>
      <option>Hey</option>
    </Select>
  ));
