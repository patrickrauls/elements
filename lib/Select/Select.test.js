import React from 'react';
import Select from './Select';
import StyledSelect from './index';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

describe('Select', () => {
  describe('unit tests', () => {
    const setup = (props, method = shallow) => method(<Select {...props} />);
    it('renders without crashing', () => {
      const wrapper = setup();
      expect(wrapper.exists()).toBe(true);
    });
    it('renders its children', () => {
      const children = [
        <option key="1" value="unique 1" />,
        <option key="2" value="unique-2" />
      ];
      const wrapper = setup({ children });
      expect(wrapper.containsAllMatchingElements(children)).toBe(true);
    });
    it('initially displays the placeholder', () => {
      const placeholder = 'Foo';
      const wrapper = setup({ placeholder });
      expect(wrapper.contains(placeholder)).toBe(true);
    });
    it('displays a given icon', () => {
      const icon = "globe";
      const wrapper = setup({ icon });
      expect(wrapper.find('.ellies-select-icon').shallow().prop('icon')).toEqual(icon);
    });
    it('displays a required Flag when dirty and required prop is true', () => {
      const wrapper = setup({ required: true }, mount);
      wrapper.find('select').simulate('focus');
      wrapper.find('select').simulate('blur');
      expect(wrapper.find('Flag').prop('type')).toEqual('required');
      wrapper.unmount();
    });
  });
  describe('interaction tests', () => {
    let wrapper;
    afterEach(() => {
      wrapper.unmount();
    });
    const setup = (props, method = mount) => method(<Select {...props} />);
    it('fires a given onChange function on change', () => {
      const onChange = jest.fn()
      wrapper = setup({ onChange });
      wrapper.find('select').simulate('change');
      expect(onChange).toHaveBeenCalled();
    });
    it('fires a given onBlur function on blur', () => {
      const onBlur = jest.fn();
      wrapper = setup({ onBlur });
      wrapper.find('select').simulate('blur');
      expect(onBlur).toHaveBeenCalled();
    });
    it('fires a given onFocus function on focus', () => {
      const onFocus = jest.fn();
      wrapper = setup({ onFocus });
      wrapper.find('select').simulate('focus');
      expect(onFocus).toHaveBeenCalled();
    });
    it('puts focus on the select when the component is clicked', () => {
      const focus = jest.fn();
      const currentTarget = {
        children: [{
          tagName: 'SELECT',
          focus
        }]
      };
      wrapper = setup(null, shallow);
      wrapper.simulate('click', { currentTarget });
      expect(focus).toHaveBeenCalled();
    });
    it('syncs its value on change', () => {
      const value = 'abc123';
      const event = {
        target: {
          value,
          checkValidity() {}
        }
      };
      wrapper = setup(null, shallow);
      wrapper.find('select').simulate('change', event);
      expect(wrapper.find('select').prop('value')).toEqual(value);
    });
  });
  describe('ui tests', () => {
    const setup = (props, method = shallow) => method(<StyledSelect {...props} />);
    it('adds styles to the Select', () => {
      const wrapper = setup();
      expect(wrapper).toMatchSnapshot();
    });
    it('puts opacity to 0.4 when disabled', () => {
      const wrapper = setup({ disabled: true });
      expect(toJson(wrapper)).toHaveStyleRule('opacity', '0.4');
    });
  });
});
