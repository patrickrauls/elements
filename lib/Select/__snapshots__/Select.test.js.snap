// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select ui tests adds styles to the Select 1`] = `
.c0 {
  position: relative;
  margin-top: 2.1875rem;
  border-bottom: 0.125rem solid #474f4f;
  width: 100%;
  background-color: rgba(0,0,0,0);
  padding: 0;
  opacity: 1;
  cursor: pointer;
}

.c0.ellies-disabled-select {
  opacity: 0.4;
}

.c0.ellies-select-focused {
  border-bottom: 0.125rem solid #021828;
  color: #021828;
}

.c0.ellies-invalid,
.c0.ellies-required {
  color: #af1a12;
  border-bottom: 0.125rem solid #af1a12;
}

.c0 select {
  cursor: pointer;
  -moz-appearance: none;
  -webkit-appearance: none;
  border: none;
  width: 100%;
  padding: 0 1.125rem 0.25rem 0;
  padding-left: 0;
  font-size: 1rem;
  background-color: rgba(0,0,0,0);
}

.c0 select.placeholder {
  color: #474f4f;
  opacity: 0.9;
}

.c0 select:focus {
  outline: none;
  color: #021828;
}

.c0 .ellies-select-icon {
  position: absolute;
  height: 1rem;
  width: 0.875rem;
  left: 0;
  bottom: 0.25rem;
}

.c0 select.invalid:not(:focus) ~ .flag {
  display: block;
  -webkit-animation: fadein 0.3s;
  animation: fadein 0.3s;
}

.c0 .select:focus {
  display: none;
}

<Select
  className="c0"
  disabled={false}
  icon=""
  name=""
  required={false}
/>
`;
